packages/node/src/indexer/mmr.service.ts

```diff
@@ -0,0 +1,159 @@
+const DEFAULT_LEAF = Buffer.from('0x000000000000000000000000000000');
+const DEFAULT_FETCH_RANGE = 100;
+
+@Injectable()
+export class MmrService implements OnApplicationShutdown {
+  private isShutdown = false;
+  private blockOffset: number;
+  private schema: string;
+  private metadataRepo: any;
+  private fileBasedMmr: MMR;
+  private poiRepo: PoiRepo;
+  private poiBlocksBuffer: BlockedQueue<ProofOfIndex>;
+  private mmrStartHeight: number;
+
+  constructor(
+    protected nodeConfig: NodeConfig,
+    protected project: SubqueryProject,
+    protected sequelize: Sequelize,
+    private poiService: PoiService,
+  ) {
+    this.poiBlocksBuffer = new BlockedQueue<ProofOfIndex>(
+      DEFAULT_FETCH_RANGE * 3,
+    );
+  }
+
+  onApplicationShutdown(): void {
+    this.isShutdown = true;
+  }
+
+  async init(schema: string): Promise<void> {
+    this.schema = schema;
+    this.metadataRepo = MetadataFactory(this.sequelize, this.schema);
+    this.poiRepo = PoiFactory(this.sequelize, this.schema);
+    this.blockOffset = await this.fetchBlockOffsetFromDb();
+    this.ensureFileBasedMmr(this.nodeConfig.mmrPath);
+    this.mmrStartHeight =
+      (await this.fileBasedMmr.getLeafLength()) + this.blockOffset;
+    logger.info(
+      `file based database MMR start with height ${this.mmrStartHeight}`,
+    );
+  }
+
+  async fetchBlockOffsetFromDb(): Promise<number | null> {
+    const blockOffset = await this.metadataRepo.findOne({
+      where: { key: 'blockOffset' },
+    });
+    if (blockOffset === null) {
+      throw new Error(`Poi service failed to fetch block offset from metadata`);
+    }
+    return Number(blockOffset.value);
+  }
+
+  async appendMmrNode(poiBlock: ProofOfIndex): Promise<void> {
+    let mmrRoot: Uint8Array;
+    const newLeaf = poiBlock.hash;
+    if (newLeaf.length !== DEFAULT_WORD_SIZE) {
+      throw new Error(
+        `Append Mmr failed, input data length should be ${DEFAULT_WORD_SIZE}`,
+      );
+    }
+    const blockLeafHeight = poiBlock.id - this.blockOffset - 1; //leaf index alway -1
+    const mmrNextLeafIndex = await this.fileBasedMmr.getLeafLength();
+    if (blockLeafHeight === mmrNextLeafIndex) {
+      //next mmr append leaf index should be same as block height
+      await this.fileBasedMmr.append(newLeaf, blockLeafHeight);
+      mmrRoot = await this.fileBasedMmr.getRoot(blockLeafHeight);
+    } else if (blockLeafHeight < mmrNextLeafIndex) {
+      //when FileBased Mmr height is ahead of Poi table, fetch mmr from fd and compare to poi
+      mmrRoot = await this.fileBasedMmr.getRoot(blockLeafHeight);
+    }
+    await this.updatePoiMmrRoot(poiBlock.id, mmrRoot);
+    this.mmrStartHeight = poiBlock.id + 1;
+  }
+
+  async updatePoiMmrRoot(id: number, mmrValue: Uint8Array) {
+    const poiBlock = await this.poiRepo.findByPk(id);
+    if (poiBlock.mmrRoot === null) {
+      poiBlock.mmrRoot = mmrValue;
+      await poiBlock.save();
+    } else if (u8aToHex(poiBlock.mmrRoot) !== u8aToHex(mmrValue)) {
```

Comment:
Need to check Uint8Array value here, without u8aTohex this report two value is not the same.

Reply:
use `u8aEq`